featurebyte/storage/webdav.py

```diff
@@ -0,0 +1,230 @@
+"""
+Webdav storage class (rclone)
+"""
+
+from typing import AsyncGenerator
+
+import re
+from http import HTTPStatus
+from pathlib import Path
+
+import httpx
+
+from featurebyte.enum import StrEnum
+from featurebyte.logging import get_logger
+from featurebyte.storage.base import Storage
+
+logger = get_logger(__name__)
+
+
+class WebdavHTTPMethods(StrEnum):
+    """
+    Webdav HTTP methods
+    """
+
+    PROPFIND = "PROPFIND"
+    MKCOL = "MKCOL"
+    COPY = "COPY"
+    MOVE = "MOVE"
+    LOCK = "LOCK"
+    UNLOCK = "UNLOCK"
+
+
+class WebdavStorage(Storage):
+    """
+    Webdav storage class
+    """
+
+    def __init__(self, base_url: str, temp: bool = False) -> None:
+        """
+        Initialize class
+
+        Parameters
+        ----------
+        base_url: str
+            Base URL of storage
+        temp: bool
+            Is temp data
+        """
+        base_url = base_url.rstrip("/")
+
+        self.client = httpx.AsyncClient()
+        self.base_url = base_url  # http://localhost:1234
+        self.temp = temp
+
+    async def __adel__(self) -> None:
+        await self.client.aclose()
+
+    async def mkdir(self, remote_path: Path) -> None:
+        """
+        Create directory in storage, similar to mkdir -p
+
+        Parameters
+        ----------
+        remote_path: Path
+            Path of remote directories to be created
+
+        Raises
+        ------
+        FileExistsError
+            File already exists on remote path
+        """
+        paths = str(remote_path).strip("/").split("/")
+        mkdir_path = "/"
+        for path in paths[:-1]:
+            mkdir_path = f"{mkdir_path}/{path}"
+            request = self.client.build_request(
+                WebdavHTTPMethods.PROPFIND, url=f"{self.base_url}/{mkdir_path}"
+            )
+            response = await self.client.send(request)
+            await response.aclose()
```

Comment:
I think aclose only needs to be called when using a streaming request with using client as context manager